SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp

```diff
@@ -0,0 +1,550 @@
+	WorkerBuildConfigAsync();
+}
+
+FString FLocalDeploymentManager::GetProjectName()
+{
+	const FString SpatialDirectory = GetSpatialOSDirectory();
+
+	FString SpatialFileName = TEXT("spatialos.json");
+	FString SpatialFileResult;
+	FFileHelper::LoadFileToString(SpatialFileResult, *FPaths::Combine(SpatialDirectory, SpatialFileName));
+
+	TSharedPtr<FJsonObject> JsonParsedSpatialFile;
+	if (!ParseJson(SpatialFileResult, JsonParsedSpatialFile))
+	{
+		UE_LOG(LogSpatialDeploymentManager, Error, TEXT("Json parsing of spatialos.json failed. Can't get project name."));
+	}
+
+	if (!JsonParsedSpatialFile->TryGetStringField(TEXT("name"), ProjectName))
+	{
+		UE_LOG(LogSpatialDeploymentManager, Error, TEXT("'name' does not exist in spatialos.json. Can't read project name."));
+	}
+
+	return ProjectName;
+}
+
+void FLocalDeploymentManager::WorkerBuildConfigAsync()
+{
+	AsyncTask(ENamedThreads::AnyBackgroundThreadNormalTask, [this]
+	{
+		FString SpatialExe = TEXT("spatial.exe");
+		FString SpatialServiceStatusArgs = TEXT("worker build build-config");
+		const FString SpatialDirectory = GetSpatialOSDirectory();
+
+		FString WorkerBuildConfigResult;
+		int32 ExitCode;
+		ExecuteAndReadOutput(*SpatialExe, *SpatialServiceStatusArgs, *SpatialDirectory, WorkerBuildConfigResult, ExitCode);
+
+		if (ExitCode == ExitCodeSuccess)
+		{
+			UE_LOG(LogSpatialDeploymentManager, Error, TEXT("Building worker configurations failed. Please ensure your .worker.json files are correct."));
+		}
+	});
+}
+
+bool FLocalDeploymentManager::ParseJson(FString RawJsonString, TSharedPtr<FJsonObject>& JsonParsed)
+{
+	TSharedRef<TJsonReader<TCHAR>> JsonReader = TJsonReaderFactory<TCHAR>::Create(RawJsonString);
+	return FJsonSerializer::Deserialize(JsonReader, JsonParsed);
+}
+
+// ExecuteAndReadOutput exists so that a spatial command window does not spawn when using 'spatial.exe'. It does not however allow reading from StdErr.
+// For other processes which do not spawn cmd windows, use ExecProcess instead.
+void FLocalDeploymentManager::ExecuteAndReadOutput(FString Executable, FString Arguments, FString DirectoryToRun, FString& OutResult, int32& ExitCode)
+{
+	UE_LOG(LogSpatialDeploymentManager, Verbose, TEXT("Executing '%s' with arguments '%s' in directory '%s'"), *Executable, *Arguments, *DirectoryToRun);
+
+	void* ReadPipe = nullptr;
+	void* WritePipe = nullptr;
+	ensure(FPlatformProcess::CreatePipe(ReadPipe, WritePipe));
+
+	FProcHandle ProcHandle = FPlatformProcess::CreateProc(*Executable, *Arguments, false, true, true, nullptr, 1 /*PriorityModifer*/, *DirectoryToRun, WritePipe);
+
+	if (ProcHandle.IsValid())
+	{
+		for (bool bProcessFinished = false; !bProcessFinished; )
+		{
+			bProcessFinished = FPlatformProcess::GetProcReturnCode(ProcHandle, &ExitCode);
+
+			OutResult = OutResult.Append(FPlatformProcess::ReadPipe(ReadPipe));
+			FPlatformProcess::Sleep(0.01f);
+		}
+
+		ProcHandle.Reset();
```

Comment:
Shouldn't we be closing the processes before this?

Reply:
Fair point. This also does Reset for you so changing the code.